### PR TITLE
General class for testing any Aggregation Implementation

### PR DESCRIPTION
**Pull request by himanshug**
and test for hyperUnique aggregation and post-aggregation

Also, found and fixed [potential bug] in HyperUniqueFinalizingPostAggregator. It did not give option to user for specifying a separate output name and wouldn't go inside a query directly, failing with the error that an aggregation with that name already exists.

### CODE REVIEW

**Comment by xvrl** [comment]
@himanshug this seems to be mainly testing groupBy aggregations. Can you explain how this is different from the existing `GroupByQueryRunnerTest`. I feel a lot of the test setup to create the index is similar to what we already do as part of `QueryRunnerTestHelper`

**Comment by himanshug** [comment]
@xvrl it is different in that it tries to exercise all the features of an aggregator implementation e.g. json serialization/deserialization of sketch object, serde of aggregator factories and all the other aspects. It simulates index creation also using the aggregator provided.
I have given hyperUnique test as an example, please how simple it is now to test the aggregator. Group-by query is only chosen as a way to exercise querying aspect only.
I don't know if I can take any other test class and do that without this new class and setup.

File: processing/src/main/java/io/druid/query/aggregation/AggregationTestHelper.java [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+                                      try {
+                                        return makeStringSerdeQueryRunner(
+                                            mapper,
+                                            toolChest,
+                                            query,
+                                            factory.createRunner(segment)
+                                        );
+                                      }
+                                      catch (Exception ex) {
+                                        Throwables.propagate(ex);
```
Comment (xvrl) [inline processing/src/main/java/io/druid/query/aggregation/AggregationTestHelper.java]: you can simply throw the result of Throwables.propagate to make the compiler happy

File: processing/src/main/java/io/druid/query/aggregation/AggregationTestHelper.java [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+          @Override
+          public QueryableIndexSegment apply(File segmentDir)
+          {
+            try {
+              return new QueryableIndexSegment("", IndexIO.loadIndex(segmentDir));
+            }
+            catch (IOException ex) {
+              Throwables.propagate(ex);
+            }
+            throw new RuntimeException("unreachable");
```
Comment (xvrl) [inline processing/src/main/java/io/druid/query/aggregation/AggregationTestHelper.java]: same here, see my comment below

File: processing/src/test/java/io/druid/query/aggregation/hyperloglog/HyperUniquesAggregationTest.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+                       + "    \"columns\": [\"timestamp\", \"market\", \"quality\", \"placement\", \"placementish\", \"index\"]"
+                       + "  }"
+                       + "}";
+
+    File f = Files.createTempDir();
+    helper.createIndex(
+        new File(this.getClass().getClassLoader().getResource("druid.sample.tsv").getFile()),
+        parseSpec,
+        metricSpec,
+        f,
```
Comment (xvrl) [inline processing/src/test/java/io/druid/query/aggregation/hyperloglog/HyperUniquesAggregationTest.java]: we can make things even easier to test if we don't have to pass a temp directory.

File: processing/src/test/java/io/druid/query/aggregation/hyperloglog/HyperUniquesAggregationTest.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+  {
+    AggregationTestHelper helper = new AggregationTestHelper(Lists.newArrayList(new AggregatorsModule()));
+
+    String metricSpec = "[{"
+                        + "\"type\": \"hyperUnique\","
+                        + "\"name\": \"index_hll\","
+                        + "\"fieldName\": \"market\""
+                        + "}]";
+
+    String parseSpec = "{"
```
Comment (xvrl) [inline processing/src/test/java/io/druid/query/aggregation/hyperloglog/HyperUniquesAggregationTest.java]: we can probably move the parseSpec and the input data into the aggregation test helper?

File: processing/src/main/java/io/druid/query/aggregation/AggregationTestHelper.java (hunk 1 above)
Comment (himanshug) [inline processing/src/main/java/io/druid/query/aggregation/AggregationTestHelper.java]: good point, why dint I think of that before :(

File: processing/src/test/java/io/druid/query/aggregation/hyperloglog/HyperUniquesAggregationTest.java (hunk 3 above)
Comment (himanshug) [inline processing/src/test/java/io/druid/query/aggregation/hyperloglog/HyperUniquesAggregationTest.java]: I wanted caller to have control on where to place the created index ,caller needs to supply it to runQueryOnSegments(..) later.

File: processing/src/test/java/io/druid/query/aggregation/hyperloglog/HyperUniquesAggregationTest.java (hunk 4 above)
Comment (himanshug) [inline processing/src/test/java/io/druid/query/aggregation/hyperloglog/HyperUniquesAggregationTest.java]: I'm not sure if I understand this. can you explain more?

File: processing/src/test/java/io/druid/query/aggregation/hyperloglog/HyperUniquesAggregationTest.java (hunk 3 above)
Comment (himanshug) [inline processing/src/test/java/io/druid/query/aggregation/hyperloglog/HyperUniquesAggregationTest.java]: thinking more, I am going to create another convenience method in helper class that wraps both createIndex and runQueryOnSegments

Comment (xvrl) [inline processing/src/test/java/io/druid/query/aggregation/hyperloglog/HyperUniquesAggregationTest.java]: why not just have createIndex return what to call runQueryOnSegments with? This make it easier to work with it and someone doesn't have to worry about creating temp dirs and cleaning them up

File: processing/src/test/java/io/druid/query/aggregation/hyperloglog/HyperUniquesAggregationTest.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+        new File(this.getClass().getClassLoader().getResource("druid.sample.tsv").getFile()),
+        parseSpec,
+        metricSpec,
+        f,
+        0,
+        QueryGranularity.NONE,
+        50000
+    );
+
+    String query = "{"
```
Comment (xvrl) [inline processing/src/test/java/io/druid/query/aggregation/hyperloglog/HyperUniquesAggregationTest.java]: Is the purpose of string queries to test serde, or would it make sense to create them programmatically to make it easier to work with?

Comment (himanshug) [inline processing/src/test/java/io/druid/query/aggregation/hyperloglog/HyperUniquesAggregationTest.java]: yes, they are intentional to test the deserialization.
helper class does have other methods to directly pass objects, but calling test should provide strings.

File: processing/src/test/java/io/druid/query/aggregation/hyperloglog/HyperUniquesAggregationTest.java (hunk 3 above)
Comment (himanshug) [inline processing/src/test/java/io/druid/query/aggregation/hyperloglog/HyperUniquesAggregationTest.java]: if createIndex(..) create tmp directory to store segment and returns it then it becomes caller's responsibility to clean it up even if caller did not really create it. That is fine too.

However I am going to create another wrapper method inside helper class that does not ask user to pass tmp file and does whole thing... something like below . let me know if that does not look Ok to you.

  public Sequence<Row> createIndexAndRunQueryOnSegment(
      File inputDataFile,
      String parserJson,
      String aggregators,
      long minTimestamp,
      QueryGranularity gran,
      int maxRowCount,
      String groupByQueryJson
  ) throws Exception
  {
..
}

**Comment by himanshug** [comment]
@xvrl changes made as per the review comments.

FYI, I don't know if you noticed but I have put AggregationTestHelper.java in "src/main/.." instead of "src/test/.." because I want to be able to use this to test another complex aggregation implementation which is a separate extension.
Another option would have been to put it in "src/test/.." and start releasing "tests" jar also to maven repository when we release druid.

**Comment by xvrl** [comment]
@himanshug yes, please move it to the test folder. If you look at druid-server module it already depends on druid-processing test jars, and we have a couple of those in maven already for those modules that require test jars (http://search.maven.org/#search%7Cga%7C1%7Cio.druid)

**Comment by himanshug** [comment]
@xvrl cool, thatz great, looks like we already publish test jar to maven repo. let me move it.

**Comment by himanshug** [comment]
@xvrl moved.

**Comment by himanshug** [comment]
@xvrl actually non-null check for name, fieldName is good and defaulting name to be same is not good idea bcoz we have validations on queries to not have an agg and post-agg be same name. I chose to default to fieldName only to have backward compatibility.

**Comment by himanshug** [comment]
@xvrl added null check for fieldName though

**Comment by himanshug** [comment]
@xvrl I think this is ready for merging unless there are further review comments.

**Pull request merged**